keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1316 @@
+At first only deployments with `.spec.strategy.rollingUpdate.maxSurge` value are expected to
+respond to this request. If the cluster doesn't support the NodeMaintenance API, kubectl will
+perform the node drain in a backwards compatible way.
+
+`kubectl cordon` and `kubectl uncordon` commands will be enhanced with a warning that will warn
+the user if a node is made un/schedulable, and it collides with an existing NodeMaintenance object.
+As a consequence the node maintenance controller will reconcile the node back to the old value.
+
+### NodeMaintenance API
+
+NodeMaintenance objects serve as an intent to remove or migrate pods from a set of nodes. We will
+include Cordon and Drain toggles to support the following states/stages of the maintenance:
+1. Planning: this is to let the users know that maintenance will be performed on a particular set
+   of nodes in the future. Configured with `.spec.stage=Idle`.
+   2. Cordon: stop accepting (scheduling) new pods. Configured with `.spec.stage=Cordon`.
+3. Drain: gives an intent to drain all selected nodes by setting an `EvacuationRequest`
+   condition with `Reason="NodeMaintenance"` on the node's pods. Configured with
+   `.spec.stage=Drain`.
+4. Drain Complete: all targeted pods have been drained from all the selected nodes. The nodes can
+   be upgraded, restarted, or shut down. The configuration is still kept at `.spec.stage=Drain` and
+   `Drained` condition is set to `"True"` on the node maintenance object.
+5. Maintenance Complete: make the nodes schedulable again once the node maintenance is done.
+   Configured with `.spec.stage=Complete`.
+
+```golang
+
+// +enum
+type NodeMaintenanceStage string
+
+const (
+// Idle does not interact with the cluster.
+Idle NodeMaintenanceStage = "Idle"
+// Cordon cordons all selected nodes by making them unschedulable.
+Cordon NodeMaintenanceStage = "Cordon"
+// Drain:
+// 1. Cordons all selected nodes by making them unschedulable.
+// 2. Gives an intent to drain all selected nodes by setting an EvacuationRequest condition on the
+//    node's pods.
+Drain NodeMaintenanceStage = "Drain"
+// Complete:
+// 1. Removes EvacuationRequest on all the pods targeted by this NodeMaintenance.
+// 2. Uncordons all selected nodes by making them schedulable again, unless there is not another
+//    maintenance in progress.
+Complete NodeMaintenanceStage = "Complete"
+)
+
+type NodeMaintenance struct {
+    ...
+    Spec NodeMaintenanceSpec
+    Status NodeMaintenanceStatus
+}
+
+type NodeMaintenanceSpec struct {
+    // +required
+    NodeSelector *v1.NodeSelector
+
+    // The order of the stages is Idle -> Cordon -> Drain -> Complete.
+    //
+    // - The Cordon or Drain stage can be skipped by setting the stage to Complete.
+    // - The NodeMaintenance object is moved to the Complete stage on deletion unless the Idle stage has been set.
+    //
+    // The default value is Idle.
+    Stage NodeMaintenanceStage
```

Comment:
Alternative (to be added later in the document): bring more details within Kubernetes' own purview.

The API defines a list (array) of stages; you can only add to the array and you can only add valid transitions.

Example transition chains:
- Upcoming → Cordon → Drain → Completed
- Upcoming → Cancelled
- Cordon → Drain → Cancelled
- Upcoming → Cordon → Drain → TerminateNode → Completed (_TerminateNode_ might be something we add for beta)

In YAML it might look like:
```yaml
…
  maintenanceStages:
    - stage: Upcoming
    - stage: Cordon
      maxStageDurationSeconds: 60
    - stage: Drain
      maxStageDurationSeconds: 1800
```

the node this affects are therefore expected to be mid-drain; the control plane adds a Completed stage after the duration limit times out.

This feels more extensible.

Reply:
I also had this idea in mind too, but I am still not sure. The stage is sort of like a status field but one that can be changed by the user or a controller mid-flight.

Maybe we can have an additional field like `maintenancePlan`? I have to think this through. Also to get it in-line with the additional features.

Reply:
We list alternatives to tell people what we've ruled out; they go there despite _not_ being our favorite approach.

Reply:
I have introduced a drainPlan to the API. And also StageStatuses which mention which tracks the stages that were started and their timestamps.